README
NAME
    Data::ModeMerge - Merge two nested data structures, with merging modes
    and options

VERSION
    version 0.30

SYNOPSIS
        use Data::ModeMerge;

        my $hash1 = { a=>1,    c=>1, d=>{  da =>[1]} };
        my $hash2 = { a=>2, "-c"=>2, d=>{"+da"=>[2]} };


        # if you want Data::ModeMerge to behave like many other merging
        # modules (e.g. Hash::Merge or Data::Merger), turn off modes
        # (prefix) parsing and options key parsing.

        my $mm = Data::ModeMerge->new(config => {parse_prefix=>0, options_key=>undef});
        my $res = $mm->merge($hash1, $hash2);
        die $res->{error} if $res->{error};
        # $res->{result} -> { a=>2, c=>1, "-c"=>2, d=>{da=>[1], "+da"=>[2]} }


        # otherwise Data::ModeMerge will parse prefix as well as options
        # key

        my $res = $mm->merge($hash1, $hash2);
        die $res->{error} if $res->{error};
        # $res->{result} -> { a=>2, c=>-1, d=>{da=>[1,2]} }

        $res = $merge({  a =>1, {  a2 =>1, ""=>{parse_prefix=>0}},
                      {".a"=>2, {".a2"=>2                       }});
        # $res->{result} -> { a=>12, {a2=>1, ".a2"=>2} }, parse_prefix is turned off in just the subhash


        # procedural interface

        my $res = mode_merge($hash1, $hash2, {allow_destroy_hash=>0});

DESCRIPTION
    There are already several modules on CPAN to do recursive data structure
    merging, like Data::Merger and Hash::Merge. "Data::ModeMerge" differs in
    that it offers merging "modes" and "options". It provides greater
    flexibility on what the result of a merge between two data should/can
    be. This module may or may not be what you need.

    One application of this module is in handling configuration. Often there
    are multiple levels of configuration, e.g. in your typical Unix
    command-line program there are system-wide config file in /etc, per-user
    config file under ~/, and command-line options. It's convenient
    programatically to load each of those in a hash and then merge
    system-wide hash with the per-user hash, and then merge the result with
    the command-line hash to get the a single hash as the final
    configuration. Your program can from there on deal with this just one
    hash instead of three.

    In a typical merging process between two hashes (left-side and
    right-side), when there is a conflicting key, then the right-side key
    will override the left-side. This is usually the desired behaviour in
    our said program as the system-wide config is there to provide defaults,
    and the per-user config (and the command-line arguments) allow a user to
    override those defaults.

    But suppose that the user wants to *unset* a certain configuration
    setting that is defined by the system-wide config? She can't do that
    unless she edits the system-wide config (in which she might need admin
    rights), or the program allows the user to disregard the system-wide
    config. The latter is usually what's implemented by many Unix programs,
    e.g. the "-noconfig" command-line option in "mplayer". But this has two
    drawbacks: a slightly added complexity in the program (need to provide a
    special, extra comand-line option) and the user loses all the default
    settings in the system-wide config. What she needed in the first place
    was to just unset *a single setting* (a single key-value pair of the
    hash).

    Data::ModeMerge comes to the rescue. It provides a so-called "DELETE
    mode".

     mode_merge({foo=>1, bar=>2}, {"!foo"=>undef, bar=>3, baz=>1});

    will result ini:

     {bar=>3, baz=>1}

    The "!" prefix tells Data::ModeMerge to do a DELETE mode merging. So the
    final result will lack the "foo" key.

    On the other hand, what if the system admin wants to *protect* a certain
    configuration setting from being overriden by the user or the
    command-line? This is useful in a hosting or other retrictive
    environment where we want to limit users' freedom to some levels. This
    is possible via the KEEP mode merging.

     mode_merge({"^bar"=>2, "^baz"=>1}, {bar=>3, "!baz"=>0, qux=>7});

    will result in:

     {"^bar"=>2, "^baz"=>1, qux=>7}

    effectively protecting "bar" and "baz" from being overriden/deleted/etc.

    Aside from the two mentioned modes, there are also a few others
    available by default: ADD (prefix "+"), CONCAT (prefix "."), SUBTRACT
    (prefix "-"), as well as the plain ol' NORMAL/override (optional prefix
    "*").

    You can add other modes by writing a mode handler module.

    You can change the default prefixes for each mode if you want. You can
    disable each mode individually.

    You can default to always using a certain mode, like the NORMAL mode,
    and ignore all the prefixes, in which case Data::ModeMerge will behave
    like most other merge modules.

    There are a few other options like whether or not the right side is
    allowed a "change the structure" of the left side (e.g. replacing a
    scalar with an array/hash, destroying an existing array/hash with
    scalar), maximum length of scalar/array/hash, etc.

    You can change default mode, prefixes, disable/enable modes, etc on a
    per-hash basis using the so-called options key. See the OPTIONS KEY
    section for more details.

    This module can handle (though not all possible cases)
    circular/recursive references.

MERGING PREFIXES AND YOUR DATA
    Merging with this module means you need to be careful when your hash
    keys might contain one of the mode prefixes characters by accident,
    because it will trigger the wrong merge mode and moreover the prefix
    characters will be stripped from the final result (unless you configure
    the module not to do so).

    A rather common case is when you have regexes in your hash keys. Regexes
    often begins with "^", which coincidentally is a prefix for the KEEP
    mode. Or perhaps you have dot filenames as hash keys, where it clashes
    with the CONCAT mode. Or perhaps shell wildcards, where "*" is also used
    as the prefix for NORMAL mode.

    To avoid clashes, you can either:

    *   exclude the keys using
        "exclude_merge"/"include_merge"/"exclude_parse"/"include_parse"
        config settings

    *   turn off some modes which you don't want via the "disable_modes"
        config

    *   change the prefix for that mode so that it doesn't clash with your
        data via the "set_prefix" config

    *   disable prefix parsing altogether via setting "parse_prefix" config
        to 0

    You can do this via the configuration, or on a per-hash basis, using the
    options key.

    See Data::ModeMerge::Config for more details on configuration.

OPTIONS KEY
    Aside from merging mode prefixes, you also need to watch out if your
    hash contains a "" (empty string) key, because by default this is the
    key used for options key.

    Options key are used to specify configuration on a per-hash basis.

    If your hash keys might contain "" keys which are not meant to be an
    options key, you can either:

    *   change the name of the key for options key, via setting
        "options_key" config to another string.

    *   turn off options key mechanism, by setting "options_key" config to
        undef.

    See Data::ModeMerge::Config for more details about options key.

MERGING MODES
  NORMAL (optional '*' prefix on left/right side)
     mode_merge({  a =>11, b=>12}, {  b =>22, c=>23}); # {a=>11, b=>22, c=>23}
     mode_merge({"*a"=>11, b=>12}, {"*b"=>22, c=>23}); # {a=>11, b=>22, c=>23}

  ADD ('+' prefix on the right side)
     mode_merge({i=>3}, {"+i"=>4, "+j"=>1}); # {i=>7, j=>1}
     mode_merge({a=>[1]}, {"+a"=>[2, 3]}); # {a=>[1, 2, 3]}

    Additive merge on hashes will be treated like a normal merge.

  CONCAT ('.' prefix on the right side)
     mode_merge({i=>3}, {".i"=>4, ".j"=>1}); # {i=>34, j=>1}

    Concative merge on arrays will be treated like additive merge.

  SUBTRACT ('-' prefix on the right side)
     mode_merge({i=>3}, {"-i"=>4}); # {i=>-1}
     mode_merge({a=>["a","b","c"]}, {"-a"=>["b"]}); # {a=>["a","c"]}

    Subtractive merge on hashes behaves like a normal merge, except that
    each key on the right-side hash without any prefix will be assumed to
    have a DELETE prefix, i.e.:

     mode_merge({h=>{a=>1, b=>1}}, {-h=>{a=>2, "+b"=>2, c=>2}})

    is equivalent to:

     mode_merge({h=>{a=>1, b=>1}}, {h=>{"!a"=>2, "+b"=>2, "!c"=>2}})

    and will merge to become:

     {h=>{b=>3}}

  DELETE ('!' prefix on the right side)
     mode_merge({x=>WHATEVER}, {"!x"=>WHATEVER}); # {}

  KEEP ('^' prefix on the left/right side)
    If you add '^' prefix on the left side, it will be protected from being
    replaced/deleted/etc.

     mode_merge({'^x'=>WHATEVER1}, {"x"=>WHATEVER2}); # {x=>WHATEVER1}

    For hashes, KEEP mode means that all keys on the left side will not be
    replaced/modified/deleted, *but* you can still add more keys from the
    right side hash.

     mode_merge({a=>1, b=>2, c=>3},
                {a=>4, '^c'=>1, d=>5},
                {default_mode=>'KEEP'});
                # {a=>1, b=>2, c=>3, d=>5}

    Multiple prefixes on the right side is allowed, where the merging will
    be done by precedence level (highest first):

     mode_merge({a=>[1,2]}, {'-a'=>[1], '+a'=>[10]}); # {a=>[2,10]}

    but not on the left side:

     mode_merge({a=>1, '^a'=>2}, {a=>3}); # error!

    Precedence levels (from highest to lowest):

     KEEP
     NORMAL
     SUBTRACT
     CONCAT ADD
     DELETE

FUNCTIONS
  mode_merge($l, $r[, $config_vars])
    A non-OO wrapper for merge() method. Exported by default. See "merge"
    method for more details.

ATTRIBUTES
  config
    A hashref for config. See Data::ModeMerge::Config.

  modes
  combine_rules
  path
  errors
  mem
  cur_mem_key
METHODS
    For typical usage, you only need merge().

  push_error($errmsg)
    Used by mode handlers to push error when doing merge. End users normally
    should not need this.

  register_mode($name_or_package_or_obj)
    Register a mode. Will die if mode with the same name already exists.

  check_prefix($hash_key)
    Check whether hash key has prefix for certain mode. Return the name of
    the mode, or undef if no prefix is detected.

  check_prefix_on_hash($hash)
    This is like "check_prefix" but performed on every key of the specified
    hash. Return true if any of the key contain a merge prefix.

  add_prefix($hash_key, $mode)
    Return hash key with added prefix with specified mode. Log merge error
    if mode is unknown or is disabled.

  remove_prefix($hash_key)
    Return hash key will any prefix removed.

  remove_prefix_on_hash($hash)
    This is like "remove_prefix" but performed on every key of the specified
    hash. Return the same hash but with prefixes removed.

  merge($l, $r)
    Merge two nested data structures. Returns the result hash: {
    success=>0|1, error=>'...', result=>..., backup=>... }. The 'error' key
    is set to contain an error message if there is an error. The merge
    result is in the 'result' key. The 'backup' key contains replaced
    elements from the original hash/array.

CREATING AND USING YOUR OWN MODE
    Let's say you want to add a mode named "FOO". It will have the prefix
    '?'.

    Create the mode handler class, e.g. "Data::ModeMerge::Mode::FOO". It's
    probably best to subclass from Data::ModeMerge::Mode::Base. The class
    must implement name(), precedence_level(), default_prefix(),
    default_prefix_re(), and
    merge_{SCALAR,ARRAY,HASH}_{SCALAR,ARRAY,HASH}(). For more details, see
    the source code of Base.pm and one of the mode handlers (e.g.
    NORMAL.pm).

    To use the mode, register it:

     my $mm = Data::ModeMerge->new;
     $mm->register_mode('FOO');

    This will require "Data::ModeMerge::Mode::FOO". After that, define the
    operations against other modes:

     # if there's FOO on the left and NORMAL on the right, what mode
     # should the merge be done in (FOO), and what the mode should be
     # after the merge? (NORMAL)
     $mm->combine_rules->{"FOO+NORMAL"} = ["FOO", "NORMAL"];

     # we don't define FOO+ADD

     $mm->combine_rules->{"FOO+KEEP"} = ["KEEP", "KEEP"];

     # and so on

FAQ
  What is this module good for? Why would I want to use this module instead of the other hash merge modules?
    If you just need to (deeply) merge two hashes, chances are you do not
    need this module. Use, for example, Hash::Merge, which is also flexible
    enough because it allows you to set merging behaviour for merging
    different types (e.g. SCALAR vs ARRAY).

    You might need this module if your data is recursive/self-referencing
    (which, last time I checked, is not handled well by Hash::Merge), or if
    you want to be able to merge differently (i.e. apply different merging
    modes) according to different prefixes on the key, or through special
    key. In other words, you specify merging modes from inside the hash
    itself.

    I originally wrote Data::ModeMerge this for Data::Schema and
    Config::Tree. I want to reuse the "parent" schema (or configuration) in
    more ways other than just override conflicting keys. I also want to be
    able to allow the parent to protect certain keys from being overriden. I
    found these two features lacking in all merging modules that I've
    evaluated prior to writing Data::ModeMerge.

SEE ALSO
    Data::ModeMerge::Config

    Other merging modules on CPAN: Data::Merger (from Data-Utilities),
    Hash::Merge, Hash::Merge::Simple

    Data::Schema and Config::Tree (among others, two modules which use
    Data::ModeMerge)

AUTHOR
    Steven Haryanto <stevenharyanto@gmail.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2013 by Steven Haryanto.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Data-ModeMerge

0.30   2013-07-16 (SHARYANTO)

       No functional changes. Replace indirect call to avoid warning from
       'indirect'.


0.29   2012-12-13 (SHARYANTO)

       Thank God the apocalypse did not happen, allowing me to do this release.

       - Fix cloning bug exposed by hash randomization of perl 5.17.6
         [RT#81807]. Turns out Data::Clone does not deep-copy objects by
         default. So now we copy configs manually.


0.28   2012-11-30 (SHARYANTO)

       - Replace Clone::Any with Data::Clone (Clone::Any fallbacks to Storable
         which doesnt support storing Regexp)

       - Add missing dependency to Test::Exception [RT#81544]


0.27   2011-02-24 (SHARYANTO)

       - regression: Clone needed as dependency


0.26   2011-02-23 (SHARYANTO)

       - switch from Mouse to Moo


0.25   2010-11-10 (SHARYANTO)

       - add use 5.010 instead of use feature (5.8.x doesn't have this yet)


0.24   2010-07-23 (SHARYANTO)

       - switch from Clone to Clone::Any (but still list Clone as dep)


0.23   2010-02-17 (SHARYANTO)

       - switch from Moose to Any::Moose (+Mouse as dependency)


0.22   2009-12-17 (SHARYANTO)

       - some work on doc
       - minor fixes


0.21   2009-12-15 (SHARYANTO)

       - replace Storable + Regexp::Copy with Clone


0.20   2009-12-15 (SHARYANTO)

       - use Regexp::Copy


0.19   2009-12-14 (SHARYANTO)

       - fix copy-paste bug


0.18   2009-12-12 (SHARYANTO)

       - bug fix: custom prefix_re got reset in hash merge

       - bug fix: restore_config() didn't recreate DM::Config object
         but hash instead


0.17   2009-12-12 (SHARYANTO)

       - INCOMPATIBLE BEHAVIOUR CHANGE: SUBTRACT merge on two hashes
         is now defined as a NORMAL merge, except that all keys on the
         right-side hash which do not have a merge prefix will be
         assumed to have a DELETE prefix. This means the SUBTRACT mode
         can now grok mode prefix and options key too.

       - INCOMPATIBLE BEHAVIOUR CHANGE: merging hash keys with DELETE
         mode now will make the key disappear even if there is no
         corresponding left or right side. merge({}, {"!a"=>1}) and
         merge({"!a"=>1}, {}) used to be {"!a"=>1} but now will be
         {}. I believe this is more intuitive/expected.


0.16   2009-12-11 (SHARYANTO)

       - some work on doc


0.15   2009-12-11 (SHARYANTO)

       - new config; premerge_pair_filter

       - NORMAL mode can handle CODE refs

       - can merge recursive/circular references (though not all possible cases)


0.14    2009-12-07 (SHARYANTO)

        - Minor fix to Makefile.PL

        - Minor fix to options key handling


0.13    2009-12-07 (SHARYANTO)

        This is a major refactoring of the module. The goal is to make
        the code more flexible (e.g. adding future modes, changing
        prefix for each mode, etc). Configuration is also
        significantly expanded.

        - rename dist: Data-PrefixMerge -> Data-ModeMerge (to reflect
          the fact that merging mode can be selected not only via a
          "prefix" but more generally by a regex test on key).

        - incompatible behaviour: all hash prefixes on the left-side
          and right-side hash will be removed recursively after merge,
          even if it is not a hash-hash merge, e.g. mode_merge({a=>1},
          {b=>{"+c"=>1}) used to become {a=>1, b=>{"+c"=>1}} but now
          it will become {a=>1, b=>{c=>1}}. this ensures that all
          prefixes are "cleaned" after merge.

        - incompatible behaviour: KEEP prefix (^) is now always sticky
          so it always protects future merges (equivalent to old
          config's preserve_keep_prefix). If you want KEEP prefix to
          be removed after merge, set config C<readd_prefix> to 0.

        - incompatible change: options key is now enabled by
          default, with "" (empty string) as the key.

        - incompatible change: configuration 'parse_hash_key_prefix'
          renamed to 'parse_prefix'.

        - removed: remove_keep_prefix(), replaced by the more generic
          remove_prefix().

        - configurable prefix for each merging mode.

        - new config: set_prefix allows you to change prefixes, even
          in the middle of recursive merge.

        - new config: allow_override & disallow_override.

        - new config: allow_create_array, allow_create_hash,
          allow_destroy_array, allow_destroy_hash.

        - new config: exclude_parse, exclude_parse_regex,
          include_parse, include_parse_regex.

        - new config: exclude_merge, exclude_merge_regex,
          include_merge, include_merge_regex.

        - new config: disable_modes. each merging mode can now be
          individually enabled/disabled through configuration.

        - options key now understands almost all configuration.


0.12    2009-11-25 (SHARYANTO)

        - new option: hash_options_key

        - fixed random ordering bug (closes #51799)


0.11    2009-11-23 (SHARYANTO)

        - add max_level parameter to remove_keep_prefixes()


0.10    2009-11-23 (SHARYANTO)

        - handle recursion in remove_keep_prefixes()


0.09    2009-11-18 (SHARYANTO)

        - fix bug in remove_keep_prefixes()


0.08    2009-11-18 (SHARYANTO)

        - new method: remove_keep_prefixes()


0.07    2009-11-18 (SHARYANTO)

        - minor fixes


0.06    2009-11-08 (SHARYANTO)

        - INCOMPATIBLE CHANGE: getting and setting config is now done via
          $merger->config->confname instead of $merger->config->{confname} to
          catch config name typos at compile time.


0.05    2009-06-22 (SHARYANTO)

        - yet another incompatible change: keep prefix now uses '^' instead of
          '!' on the left side. this allows keep mode on the right side. on a
          series of merge we can now introduce keep mode at one point (not just
          on the first hash) and then preserve that keep mode afterwards.



lib/Data/ModeMerge/Config.pm
package Data::ModeMerge::Config;

use 5.010;
use Moo;

our $VERSION = '0.30'; # VERSION

has recurse_hash          => (is => 'rw', default => sub{1});
has recurse_array         => (is => 'rw', default => sub{0});
has parse_prefix          => (is => 'rw', default => sub{1});
has wanted_path           => (is => 'rw');
has default_mode          => (is => 'rw', default => sub{'NORMAL'});
has disable_modes         => (is => 'rw');
has allow_create_array    => (is => 'rw', default => sub{1});
has allow_create_hash     => (is => 'rw', default => sub{1});
has allow_destroy_array   => (is => 'rw', default => sub{1});
has allow_destroy_hash    => (is => 'rw', default => sub{1});
has exclude_parse         => (is => 'rw');
has exclude_parse_regex   => (is => 'rw');
has include_parse         => (is => 'rw');
has include_parse_regex   => (is => 'rw');
has exclude_merge         => (is => 'rw');
has exclude_merge_regex   => (is => 'rw');
has include_merge         => (is => 'rw');
has include_merge_regex   => (is => 'rw');
has set_prefix            => (is => 'rw');
has readd_prefix          => (is => 'rw', default => sub{1});
has premerge_pair_filter  => (is => 'rw');
has options_key           => (is => 'rw', default => sub{''});
has allow_override        => (is => 'rw');
has disallow_override     => (is => 'rw');

# list of config settings only available in merger-object's config
# (not in options key)
sub _config_config {
    state $a = [qw/
        wanted_path
        options_key
        allow_override
        disallow_override
                  /];
}

# list of config settings available in options key
sub _config_ok {
    state $a = [qw/
        recurse_hash
        recurse_array
        parse_prefix
        default_mode
        disable_modes
        allow_create_array
        allow_create_hash
        allow_destroy_array
        allow_destroy_hash
        exclude_parse
        exclude_parse_regex
        include_parse
        include_parse_regex
        exclude_merge
        exclude_merge_regex
        include_merge
        include_merge_regex
        set_prefix
        readd_prefix
        premerge_pair_filter
                  /];
}

1;
# ABSTRACT: Data::ModeMerge configuration


__END__
=pod

=head1 NAME

Data::ModeMerge::Config - Data::ModeMerge configuration

=head1 VERSION

version 0.30

=head1 SYNOPSIS

 # getting configuration
 if ($mm->config->allow_extra_hash_keys) { ... }

 # setting configuration
 $mm->config->max_warnings(100);

=head1 DESCRIPTION

Configuration variables for Data::ModeMerge.

=head1 ATTRIBUTES

=head2 recurse_hash => BOOL

Context: config, options key

Default: 1

Whether to recursively merge hash. When 1, each key-value pair between
2 hashes will be recursively merged. Otherwise, the right-side hash
will just replace the left-side.

Options key will not be parsed under recurse_hash=0.

Example:

 mode_merge({h=>{a=>1}}, {h=>{b=>1}}                   ); # {h=>{a=>1, b=>1}}
 mode_merge({h=>{a=>1}}, {h=>{b=>1}}, {recurse_hash=>0}); # {h=>{b=>1}}

=head2 recurse_array => BOOL

Context: config, options key

Default: 0

Whether to recursively merge array. When 1, each element is
recursively merged. Otherwise, the right-side array will just replace
the left-side.

Example:

 mode_merge([1, 1], [4]                    ); # [4, 1]
 mode_merge([1, 1], [4], {recurse_array=>0}); # [2]

=head2 parse_prefix => BOOL

Context: config, options key

Default: 1

Whether to parse merge prefix in hash keys. If set to 0, merging
behaviour is similar to most other nested merge modules.

 mode_merge({a=>1}, {"+a"=>2}                   ); # {a=>3}
 mode_merge({a=>1}, {"+a"=>2}, {parse_prefix=>0}); # {a=>1, "+a"=>2}

=head2 wanted_path => ARRAYREF

Context: config, options key

Default: undef

If set, merging is only done to the specified "branch". Useful to save
time/storage when merging large hash "trees" while you only want a
certain branch of the trees (e.g. resolving just a config variable
from several config hashes).

Example:

 mode_merge(
   {
    user => {
      steven => { quota => 100, admin => 1 },
      tommie => { quota =>  50, admin => 0 },
      jimmy  => { quota => 150, admin => 0 },
    },
    groups => [qw/admin staff/],
   },
   {
    user => {
      steven => { quota => 1000 },
    }
   }
 );

With wanted_path unset, the result would be:

   {
    user => {
      steven => { quota => 1000, admin => 1 },
      tommie => { quota =>   50, admin => 0 },
      jimmy  => { quota =>  150, admin => 0 },
    }
    groups => [qw/admin staff/],
   }

With wanted_path set to ["user", "steven", "quota"] (in other words,
you're saying that you'll be disregarding other branches), the result
would be:

   {
    user => {
      steven => { quota => 1000, admin => undef },
    }
   }

=head2 default_mode => 'NORMAL' | 'ADD' | 'CONCAT' | 'SUBTRACT' | 'DELETE' | 'KEEP' | ...

Context: config, options key

Default: NORMAL

Example:

 mode_merge(3, 4                         ); # 4
 mode_merge(3, 4, {default_mode => "ADD"}); # 7

=head2 disable_modes => ARRAYREF

Context: config, options key

Default: []

List of modes to ignore the prefixes of.

Example:

 mode_merge({add=>1, del=>2, concat=>3},
            {add=>2, "!del"=>0, .concat=>4},
            {disable_modes=>[qw/CONCAT/]});
 #          {add=>3,         concat=>3, .concat=>4}

See also: C<parse_prefix> which if set to 0 will in effect disable all
modes except the default mode.

=head2 allow_create_array => BOOL

Context: config, options key

Default: 1

If enabled, then array creation will be allowed (from something
non-array, like a hash/scalar). Setting to 0 is useful if you want to
avoid the merge to "change the structure" of the left side.

Example:

 mode_merge(1, [1,2]                         ); # success, result=[1,2]
 mode_merge(1, [1,2], {allow_create_array=>0}); # failed, can't create array

=head2 allow_create_hash => BOOL

Context: config, options key

Default: 1

If enabled, then hash creation will be allowed (from something
non-hash, like array/scalar). Setting to 0 is useful if you want to
avoid the merge to "change the structure" of the left side.

Example:

 mode_merge(1, {a=>1}                        ); # success, result={a=>1}
 mode_merge(1, {a=>1}, {allow_create_hash=>0}); # failed, can't create hash

=head2 allow_destroy_array => BOOL

Context: config, options key

Default: 1

If enabled, then replacing array on the left side with non-array
(e.g. hash/scalar) on the right side is allowed. Setting to 0 is
useful if you want to avoid the merge to "change the structure" of the
left side.

Example:

 mode_merge([1,2], {}                          ); # success, result={}
 mode_merge([1,2], {}, {allow_destroy_array=>0}); # failed, can't destroy array

=head2 allow_destroy_hash => BOOL

Context: config, options key

Default: 1

If enabled, then replacing hash on the left side with non-hash
(e.g. array/scalar) on the right side is allowed. Setting to 0 is
useful if you want to avoid the merge to "change the structure" of the
left side.

Example:

 mode_merge({a=>1}, []                         ); # success, result=[]
 mode_merge({a=>1}, [], {allow_destroy_hash=>0}); # failed, can't destroy hash

=head2 exclude_parse => ARRAYREF

Context: config, options key

Default: undef

The list of hash keys that should not be parsed for prefix and merged
as-is using the default mode.

If C<include_parse> is also mentioned then only keys in
C<include_parse> and not in C<exclude_parse> will be parsed for
prefix.

Example:

 mode_merge({a=>1, b=>2}, {"+a"=>3, "+b"=>4}, {exclude_parse=>["+b"]}); # {a=>4, b=>2, "+b"=>4}

=head2 exclude_parse_regex => REGEX

Context: config, options key

Default: undef

Just like C<exclude_parse> but using regex instead of list.

=head2 include_parse => ARRAYREF

Context: config, options key

Default: undef

If specified, then only hash keys listed by this setting will be
parsed for prefix. The rest of the keys will not be parsed and merged
as-is using the default mode.

If C<exclude_parse> is also mentioned then only keys in
C<include_parse> and not in C<exclude_parse> will be parsed for
prefix.

Example:

 mode_merge({a=>1, b=>2, c=>3}, {"+a"=>4, "+b"=>5, "+c"=>6},
            {include_parse=>["+a"]}); # {a=>1, "+a"=>4, b=>7, c=>3, "+c"=>6}

=head2 include_parse_regex => REGEX

Context: config, options key

Default: undef

Just like C<include_parse> but using regex instead of list.

=head2 exclude_merge => ARRAYREF

Context: config, options key

Default: undef

The list of hash keys on the left side that should not be merged and
instead copied directly to the result. All merging keys on the right
side will be ignored.

If C<include_merge> is also mentioned then only keys in
C<include_merge> and not in C<exclude_merge> will be merged.

Example:

 mode_merge({a=>1}, {"+a"=>20, "-a"=>30}, {exclude_merge=>["a"]}); # {a=>1}

=head2 exclude_merge_regex => REGEX

Context: config, options key

Default: undef

Just like C<exclude_merge> but using regex instead of list.

=head2 include_merge => ARRAYREF

Context: config, options key

Default: undef

If specified, then only hash keys listed by this setting will be
merged.

If C<exclude_merge> is also mentioned then only keys in
C<include_merge> and not in C<exclude_merge> will be merged.

Example:

 mode_merge({a=>1, b=>2, c=>3}, {"+a"=>40, "+b"=>50, "+c"=>60, "!c"=>70},
            {include_merge=>["a"]}); # {a=>41, b=>2, c=>3}

=head2 include_merge_regex => ARRAYREF

Context: config, options key

Default: undef

Just like C<include_merge> but using regex instead of list.

=head2 set_prefix => HASHREF

Context: config, options key

Default: undef

Temporarily change the prefix character for each mode. Value is
hashref where each hash key is mode and the value is a new prefix
string.

 mode_merge({a=>1, c=>2}, {'+a'=>10, '.c'=>20});                                        # {a=>11, c=>220}
 mode_merge({a=>1, c=>2}, {'+a'=>10, '.c'=>20}, {set_prefix=>{ADD=>'.', CONCAT=>'+'}}); # {a=>110, c=>22}

=head2 readd_prefix => BOOL

Context: config, options key

Default: 1

When merging two hashes, the prefixes are first stripped before
merging. After merging is done, the prefixes by default will be
re-added. This is done so that modes which are "sticky" (like KEEP)
can propagate their mode). Setting C<readd_prefix> to 0 will prevent
their stickiness.

 mode_merge({"^a"=>1}, {a=>2});                    # {"^a"=>1}
 mode_merge({"^a"=>1}, {a=>2}, {readd_prefix=>0}); # { "a"=>1}

=head2 premerge_pair_filter => CODEREF

Context: config, options key

Default: undef

Pass the key and value of each hash pair to a subroutine before
merging (and before the keys are stripped for mode prefixes). Will
push error if there is conflicting key in the hash.

The subroutine should return a list of new key(s) and value(s). If key
is undef then it means the pair should be discarded. This way, the
filter is able to add or remove pairs from the hash.

 mode_merge({a=>1}, {"+amok"=>2},
            {premerge_pair_filter=>sub{ uc(substr($_[0],0,2)), $_[1]*2 }});
 # {"A"=>6}

=head2 options_key => STR

Context: config

Default: '' (empty string)

If defined, then when merging two hashes, this key will be searched
first on the left-side and right-side hash. The values will then be
merged and override (many of) the configuration.

Options key is analogous to Apache's C<.htaccess> mechanism, which
allows setting configuration on a per-directory (per-hash)
basis. There's even an C<allow_override> config similar to Apache
directive of the same name.

If you want to disable processing of options key, set this to undef.

Example:

 mode_merge({a=>1, {x=>3}},
            {a=>2, {x=>4}},
            {default_mode=>'ADD'}); # {a=>3, {x=>7}}
 mode_merge({a=>1, {x=>3}},
            {a=>2, {x=>4, ''=>{default_mode=>'CONCAT'}}},
            {default_mode=>'ADD'}); # {a=>3, {x=>34}}

On the above example, C<default_mode> is set to ADD. But in the
{x=>...} subhash, C<default_mode> is changed to CONCAT by the options
key.

=head2 allow_override => REGEX

Context: config

Default: undef

If defined, then only config names matching regex will be able to be
set in options key.

If C<disallow_override> is also set, then only config names matching
C<allow_override> and not matching C<disallow_override> will be able
to be set in options key.

=head2 disallow_override => REGEX

Context: config

Default: undef

If defined, then config names matching regex will not be able to be
set in options key.

For example, if you want to restrict "structural changes" in merging
while still allowing options key, you can set C<allow_create_hash>,
C<allow_destroy_hash>, C<allow_create_array>, and
C<allow_destroy_array> all to 0 and C<disallow_override> to
C<allow_create|allow_destroy> to forbid overriding via options key.

If C<disallow_override> is also set, then only config names matching
C<allow_override> and not matching C<disallow_override> will be able
to be set in options key.

=head1 AUTHOR

Steven Haryanto <stevenharyanto@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Steven Haryanto.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut

